### PR TITLE
LXDProfile - Modification Status

### PR DESCRIPTION
**Pull request by SimonRichardson**
## Description of change

The follow adds a modification status, that allows us to display
when a change to a running machine fails, but we want the error
to be displayed to the operator without it disappearing.

So in the current implementation the status output uses last write
wins. Which has been good enough for most implementations up until
now, where by juju only catered for errors during provisioning.
The change here is a fundamental change, where we state that changes
during a machine update can affect what the operator sees in juju
status.

This was discussed with @jameinel for a possible solution. With
this solution we can now factor in new potential changes that happen
and can bubble them up to the operator with good effect.

The naming of the status "modification" isn't stuck in stone, but
it's to imply the outcome of an action to a machine. This can be
changed and open to suggestions.

I've had to bump the ProvisionerAPI to v8, so when looking at this
PR it's advised to look over that.

## QA steps

```sh
juju bootstrap lxd 
juju deploy ~/bundle-no-sub.yaml 
watch --color -n 1 juju status --color
```

Check that the error is kept when the machine `0` can't apply the
profile.

```
Model    Controller  Cloud/Region  Version  SLA          Timestamp
default  yyy         lxd/default   2.5.1.1  unsupported  17:30:12Z

App          Version  Status  Scale  Charm        Store  Rev  OS      Notes
lxd-profile  18.04    active      5  lxd-profile  local    0  ubuntu
ubuntu       18.04    active      5  ubuntu-lite  local    0  ubuntu

Unit            Workload  Agent  Machine  Public address  Ports  Message
lxd-profile/0*  active    idle   0        10.178.104.242         load: 5.50, 2.77, 3.26
lxd-profile/1   active    idle   1        10.178.104.171         load: 1.97, 2.22, 3.24
lxd-profile/2   active    idle   2        10.178.104.76          load: 1.97, 2.22, 3.24
lxd-profile/3   active    idle   3        10.178.104.55          load: 1.35, 1.97, 3.05
lxd-profile/4   active    idle   4        10.178.104.132         load: 1.35, 1.97, 3.05
ubuntu/0*       active    idle   0        10.178.104.242         load: 1.97, 2.22, 3.24
ubuntu/1        active    idle   1        10.178.104.171         load: 1.97, 2.22, 3.24
ubuntu/2        active    idle   2        10.178.104.76          load: 2.43, 2.31, 3.26
ubuntu/3        active    idle   3        10.178.104.55          load: 1.88, 2.21, 3.24
ubuntu/4        active    idle   4        10.178.104.132         load: 1.88, 2.21, 3.24

Machine  State    DNS             Inst id        Series  AZ  Message
0        error    10.178.104.242  juju-ad601b-0  bionic      cannot upgrade machine's lxd profile: Failed to add mount for device: Failed to run: /snap/lxd/current/bin/lxd forkmount lxc-mount juju-ad601b-0 /var/snap/lxd/common/lxd/containers /var/snap/lxd/common/lxd/logs/juju-ad601b-0/lxc.conf /var/snap/lxd/common/lxd/devices/juju-ad601b-0/unix.tun.dev-net-tun /dev/net/tun none 4096:
1        started  10.178.104.171  juju-ad601b-1  bionic      Running
2        started  10.178.104.76   juju-ad601b-2  bionic      Running
3        started  10.178.104.55   juju-ad601b-3  bionic      Running
4        started  10.178.104.132  juju-ad601b-4  bionic      Running
```

bundle-no-sub.yaml
```yaml
series: bionic
applications:
  lxd-profile:
    charm: ~/go/src/github.com/juju/juju/testcharms/charm-repo/quantal/lxd-profile
    num_units: 9
    to:
    - "0"
    - "1"
    - "2"
    - "3"
    - "4"
    - "5"
    - "6"
    - "7"
    - "8"
  ubuntu:
    charm: cs:~jameinel/ubuntu-lite
    num_units: 9
    to:
    - "0"
    - "1"
    - "2"
    - "3"
    - "4"
    - "5"
    - "6"
    - "7"
    - "8"
machines:
  "0": {}
  "1": {}
  "2": {}
  "3": {}
  "4": {}
  "5": {}
  "6": {}
  "7": {}
  "8": {}
```

### CODE REVIEW

**Comment by SimonRichardson** [comment]
!!build!!

- snapcraft io/timeout

File: state/migration_export.go [hunk 1]
```diff
@@ -1684,7 +1691,7 @@ func (e *exporter) checkUnexportedValues() error {
 	}
 
 	for key := range e.statusHistory {
-		if !e.cfg.SkipInstanceData && !strings.HasSuffix(key, "#instance") {
+		if !e.cfg.SkipInstanceData && (!strings.HasSuffix(key, "#instance") && !strings.HasSuffix(key, "#modification")) {
```
Comment (jameinel) [inline state/migration_export.go]: I feel like this should be 
!(instance || modification)

File: cmd/juju/status/status_internal_test.go [hunk 2]
```diff
@@ -4781,6 +4904,14 @@ func (s *StatusSuite) prepareTabularData(c *gc.C) *context {
 		addRemoteApplication{name: "remote-wordpress", charm: "wordpress", endpoints: []string{"db"}, isConsumerProxy: true},
 		relateApplications{"remote-wordpress", "mysql", ""},
 		addOfferConnection{sourceModelUUID: coretesting.ModelTag.Id(), name: "hosted-mysql", username: "fred", relationKey: "remote-wordpress:db mysql:server"},
+
+		// test modification status
+		addMachine{machineId: "4", job: state.JobHostUnits},
+		setAddresses{"4", network.NewAddresses("10.0.3.1")},
+		startAliveMachine{"4"},
+		setMachineStatus{"4", status.Started, ""},
+		setMachineInstanceStatus{"4", status.Started, "I am number three"},
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply cmd/juju/status/status_internal_test.go]: yeah, I'll update it as it is confusing, but that was my intention.

File: state/migration_export.go (hunk 1 above)
Comment (SimonRichardson) [reply state/migration_export.go]: Yeah, that is more readable.

File: cmd/juju/machine/package_test.go [hunk 3]
```diff
@@ -14,3 +15,11 @@ import (
 func TestPackage(t *testing.T) {
 	gc.TestingT(t)
 }
+
+func marshalStringAsJSON(str string) (interface{}, error) {
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply cmd/juju/machine/package_test.go]: Agreed.

File: apiserver/facades/agent/provisioner/provisioner.go [hunk 4]
```diff
@@ -1432,6 +1433,98 @@ func (p *ProvisionerAPI) SetInstanceStatus(args params.SetStatus) (params.ErrorR
...
+	}
+	return result, nil
+}
+
+func (p *ProvisionerAPI) setOneModificationStatus(canAccess common.AuthFunc, arg params.EntityStatusArgs) error {
+	logger.Tracef("SetInstanceStatus called with: %#v", arg)
+	mTag, err := names.ParseMachineTag(arg.Tag)
+	if err != nil {
+		logger.Warningf("SetModificationStatus called with %q which is not a valid machine tag: %v", arg.Tag, err)
+		return common.ErrPerm
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply apiserver/facades/agent/provisioner/provisioner.go]: That was my intention, but I can easily lower it.

File: apiserver/facades/agent/provisioner/provisioner_test.go [hunk 5]
```diff
@@ -437,6 +437,59 @@ func (s *withoutControllerSuite) TestSetInstanceStatus(c *gc.C) {
...
+		Message: "not really",
+		Since:   &now,
+	}
+	err = s.machines[2].SetModificationStatus(sInfo)
+	c.Assert(err, jc.ErrorIsNil)
+
+	args := params.SetStatus{
+		Entities: []params.EntityStatusArgs{
+			{Tag: s.machines[0].Tag().String(), Status: status.Pending.String(), Info: "not really",
+				Data: map[string]interface{}{"foo": "bar"}},
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply apiserver/facades/agent/provisioner/provisioner_test.go]: Sure, I'll update it.

**Comment by hmlanigan** [comment]
If I add a unit (ubuntu) with no machine modifications, its machine's modification status is left as pending:

$ juju machines --format yaml
model: default
machines:
...
  "6":
...
    machine-status:
      current: running
      message: Running
      since: 13 Feb 2019 10:28:44-05:00
    modification-status:
      current: pending
      since: 13 Feb 2019 10:28:27-05:00

**Comment by hmlanigan** [comment]
I also found a machine with a profile applied that stayed in pending, when it should be running as no error (machine 2), machine 6 is the no profile unit machine from above.
https://pastebin.canonical.com/p/fNKC8mbw6W/

**Comment by SimonRichardson** [comment]
!!build!!

**Comment by hmlanigan** [comment]
question: I remove the lxd-profile application, thus all units, the machines still have ubuntu units on them.  However the machine is still in an error state related to the profile.  Should the error be cleared?

**Comment by SimonRichardson** [comment]
Yeah it should, not sure where to do that, in destroyX probably.

File: apiserver/facades/agent/provisioner/provisioner.go [hunk 6]
```diff
@@ -1432,6 +1433,97 @@ func (p *ProvisionerAPI) SetInstanceStatus(args params.SetStatus) (params.ErrorR
...
+	}
+	machine, err := p.getMachine(canAccess, mTag)
+	if err != nil {
+		logger.Debugf("SetModificationStatus unable to get machine %q", mTag)
+		return err
+	}
+
+	// We can use the controller timestamp to get now.
+	since, err := p.st.ControllerTimestamp()
```
Comment (hmlanigan) [inline apiserver/facades/agent/provisioner/provisioner.go]: weren't we going to change setOneInstanceStatus call to time.Now() to be the same as here?

Comment (SimonRichardson) [reply apiserver/facades/agent/provisioner/provisioner.go]: yes, you're right

**Comment by SimonRichardson** [comment]
$$merge$$

**Comment by SimonRichardson** [comment]
$$merge$$

> worker_test.go:939: workerSuite.TestRemovePrimaryValidSecondaries

**Pull request merged**